designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+
+#### Hidden `generation` changes
+Although it covers most of the scenarios, this solution still leaves in edge case unsolved. While `generation` can only be increased for a certain object, by changing a child object reference to a different object we can make the aggregated generation go down, or even worse, stay the same. The latter is bad since we will identify that reconcile request as a no-op, hence no reacting to the changes. This can happen if a referenced gets changed to child object with a $generation = prevObjGeneration -1$. That decreased will be compensated by the increase in the `Cluster`'s generation.
+
+The most straight forward solution is to keep track of two generations: the `Cluster` 's generation and the aggregated generation for the child objects. 
+* If any reference changes, the  `Cluster` 's generation will increase.
+* If any child object changes, the aggregated child object generation will increase. Note that multiple child objects changing at once won't be a problem since their generation can only increase.
+
+So now the logic becomes:
+
+```mermaid
+stateDiagram-v2
+    [*] --> calculate
+
+	calculate: aggregatedGeneration = sum(child.generation)
+    reconcile: Reconcile cluster
+    observedGeneration: generation == observedGeneration and aggregatedGeneration == childrenObservedGeneration
+    updateGeneration: observedGeneration = generation
+    updateChild: childrenObservedGeneration = aggregatedGeneration
+    done: Reconciliation done?
+    calculate --> observedGeneration
+    state observedGeneration <<choice>>
+        observedGeneration --> reconcile: no
+        observedGeneration --> [*]: yes
+
+    reconcile --> done
+    state done <<choice>>
+        done --> updateGeneration: yes
+
+    updateGeneration  --> updateChild
+    updateChild  --> [*]
+```
+
+### API changes
+We propose adding two fields to the `Cluster` status:
+
+```go
+type ClusterStatus struct {
+	...
+	// observedGeneration represents the .metadata.generation the last time the
+	// cluster was successfully reconciled.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// childrenObservedGeneration represents the sum of all cluster's linked
+	// objects's .metadata.generation the last time the cluster was successfully reconciled.
+	ChildrenObservedGeneration int64 `json:"childrenObservedGeneration,omitempty"`
+}
+```
+
+### Version skew
+Although it's possible today to continue upgrading a management cluster without updating the workload cluster's `Bundles` (the versions of all components in a EKS-A cluster), we lack a maximum skew version between management and workload clusters.
+
+Since this proposal makes the isolation between management and workload clusters even bigger, we propose to add to our documentation a maximum skew version support of one minor version. This means the management cluster can only be one EKS-A minor version after the workload clusters.
```

Comment:
why does the proposal requires "the management cluster can only be one EKS-A minor version after the workload clusters."?